### ISSUE TITLE
Project autosave with unliked files leads to lost project

### ISSUE THREAD

**Issue by rpdelaney** [jliljebl/flowblade#891]
## Info

* Flowblade version: 2.6
* MLT version (Help->Runtime Environment): 6.20.0 (it sure would be nice if I could copy paste info from here btw)
* Your distribution (Ubuntu, Debian, Mint etc.): arch

## Repro

1. Open a project with flowblade and work on it for a little while
1. Get a crash, or PC loses power etc
1. Open flowblade, restore from autosave, no big deal
1. Do some more work, go to sleep
1. Repeat previous two steps once or twice again
1. Save project, close flowblade
1. In the filesystem, rename or delete one of the media files in the project
1. Open flowblade again. Get a dialog asking to relink media files. Give it the new paths
1. Get a dialogue asking if I want to restore one of the multiple autosaves. One of them is from ~15 minutes ago, but since I saved manually just before I closed, I shouldn't need to restore it, right?
1. Wrong, I lost 2 days of work.

I'm not sure if this is a bug or intended behavior, but it is not intuitive.

**Comment by jliljebl** [comment]
Losing work product for 2 days is never intended result, so let's see if I can find the underlying problem. 

I have some additional questions, the sequence of events is not completely clear to me.

* Did the 2 days data loss represent time between 1-6 because you kept the application open for 2 days and then saved at the end of this lengthy session?
* After step 8 did you select "Save Relinked Project As..." or "Close"? (*If you selected "Close" your relinking work is lost,  "Close" does not mean that relinking work is saved with original project file.*)
* Did you  ever manage to open the file from step 6 as a relinked version?
* Was there a new crash between 8 and 9?

**Comment by rpdelaney** [comment]
> Did the 2 days data loss represent time between 1-6 because you kept the application open for 2 days and then saved at the end of this lengthy session?

Close but not quite. I was saving manually, periodically, as I went. Also after my PC crashed it seemed to restore from an autosave well enough.

I am stabbing in the dark here (as I have not looked a the code at all) but I am suspicious that after restoring from an autosave, whenever I "saved" again, I was somehow saving back to the autosave and not the original project file. That would explain why it is that the first time I declined to restore from an autosave, I was met with a days-old project.

> After step 8 did you select "Save Relinked Project As..." or "Close"? (If you selected "Close" your relinking work is lost, "Close" does not mean that relinking work is saved with original project file.)

I saved the relinked project. I now have two project files, one of them with a linking error and one without, but both are old.

> Did you ever manage to open the file from step 6 as a relinked version?

Yes

> Was there a new crash between 8 and 9?

No

**Comment by jliljebl** [comment]
Ok, thanks, with this info I can try to reproduce this in the coming weeks. There might be some additional questions if I don't succeed in reproducing this.

**Comment by rpdelaney** [comment]
I may have found a simpler repro.

1. Editing video in flowblade, add a fade-in transition
1. Flowblade itself crashes while rendering the fade (my PC is fine)
1. Start flowblade, load the autosave. But I'm suspicious now
1. Do a tiny bit of work, manually auto-save, and quit flowblade via the menu
1. Restart flowblade, I am prompted again if I want to reload an auto-save

This seems to continue until I decline to load the auto-save, but any work I've done until then will be lost when I start flowblade and click "continue with default"

**Comment by jliljebl** [comment]
Great, thanks. I suspected that the relink action should not be a factor.

For the time being, when recovering with autosave use "Save As.." and give some running number for project file.

**Comment by rpdelaney** [comment]
Yes, I am using my own incremental backup solution now, as I should have from the beginning.

You probably understood what I meant, but I edited my previous comment. On step 4 I had written "manually auto-save" but this should be "manually save"

**Comment by rpdelaney** [comment]
It is possible that this bug could be isolated to "Ignore / clean-up auto-saves when the parent project was saved more recently than the auto-save was created"

**Comment by jliljebl** [comment]
> It is possible that this bug could be isolated to "Ignore / clean-up auto-saves when the parent project was saved more recently than the auto-save was created"

Quite possible. I just looked at saving related stuff and project files have an attribute **last_save_path** which is involved in a few too many things (see below) creating a complex net of possible failure modes. Things have gone a bit sideways here as there has been an unnoticed complexity build up making finding and fixing this non-trivial.

There is very likely going to be a 2.6.1 point release in late August, I'll try to fix this for that.

```bash
Flowblade/projectaction.py
193:        # If project file is moved since last save we need to update last_save_path property and save to get everything working as expected.
194:        if self.filename != editorstate.project.last_save_path:
195:            print("Project file moved since last save, save with updated last_save_path data.")
196:            editorstate.project.last_save_path = self.filename
207:            project.last_save_path = None # This gets set to the temp file saved to change profile which is not correct.
423:            if editorstate.PROJECT().last_save_path != None:
424:                persistance.save_project(editorstate.PROJECT(), editorstate.PROJECT().last_save_path)
446:    if PROJECT().last_save_path == None:
456:        persistance.save_project(PROJECT(), PROJECT().last_save_path) #<----- HERE
465:    editorpersistance.add_recent_project_path(PROJECT().last_save_path)
467:    PROJECT().events.append(projectdata.ProjectEvent(projectdata.EVENT_SAVED, PROJECT().last_save_path))
479:    if  PROJECT().last_save_path != None:
480:        open_dir = os.path.dirname(PROJECT().last_save_path)
493:        PROJECT().last_save_path = filenames[0]
499:            persistance.save_project(PROJECT(), PROJECT().last_save_path) #<----- HERE
510:            p_event = projectdata.ProjectEvent(projectdata.EVENT_CREATED_BY_SAVING, PROJECT().last_save_path)
513:            p_event = projectdata.ProjectEvent(projectdata.EVENT_SAVED_AS, (PROJECT().name, PROJECT().last_save_path))
531:        editorpersistance.add_recent_project_path(PROJECT().last_save_path)
753:        if editorstate.PROJECT().last_save_path != None:
754:            persistance.save_project(editorstate.PROJECT(), editorstate.PROJECT().last_save_path)

Flowblade/projectdata.py
89:        self.last_save_path = None

Flowblade/jobs.py
796:        if PROJECT().last_save_path != None:
797:            save_path = PROJECT().last_save_path 

Flowblade/medialinker.py
493:    if  target_project.last_save_path != None:
494:        open_dir = os.path.dirname(target_project.last_save_path)
509:        target_project.last_save_path = filenames[0]
514:            filehandle = open( target_project.last_save_path, 'w' )
525:        persistance.save_project(target_project, target_project.last_save_path)

Flowblade/app.py
1035:            if editorstate.PROJECT().last_save_path != None:
1036:                persistance.save_project(editorstate.PROJECT(), editorstate.PROJECT().last_save_path)
```

**Comment by rpdelaney** [comment]
This is a big problem.

1. Do about 4 hours work, saving as I go
2. PC locks up rendering a big clip using ffmpeg in the terminal (crash not caused by flowblade)
3. Hard reset PC
4. I am met with the dialog asking if I want to restore an autosave from 17 minutes ago. I _know_ I saved more recently than that, so I continue to the default project.
5. The project is 4 hours old.

At this point, I'm committed to using flowblade, but I don't think I can continue to work on this project until this is addressed. Even if it just didn't delete autosaves but marked them as "ignored" somehow would be a huge improvement.

**Comment by jliljebl** [comment]
I'll start doing patches for 6.2.1 bug fixes next week, I'll post more info once I have some results.

I'm quite confident we will get this fixed for that release as the pattern you have documented is quite clear.

**Comment by jliljebl** [comment]
Confirmed, there are things wrong here, we are not deleting old autosaves correctly and *Control + S* does save into old auto save.

This will get fixed for 2.6.1, this is quite bad.

**Comment by jliljebl** [comment]
This is now fixed in 2.6.1 release branch and will fix be ported to master after that release.

The problem was as suspected above, Control + S saved to autosave files which with some use patterns could lead to data loss.

Thank you for reporting. The bug was introduced when fixing another bug and did not make its self apparent with my use pattern that were to always do "Save As..." after aoto save load.
